### PR TITLE
141 user shouldnt be able to be friend with themselves

### PR DESCRIPTION
**Pull request by matboivin**
Closes: #141 , #174 

- Refacto `src/nextjs-front/context/relationship/RelationshipProvider.tsx`: remove current user since the start
- Massive refacto of `src/nextjs-front/components/FriendsTable.tsx` to avoid side effects and update user relationships
- Update design of `friends` page

Screenshot at 2022-05-18 19-25-03

### CODE REVIEW

File: src/nestjs-back/src/users/users.controller.ts [hunk 1]
```diff
@@ -41,17 +41,6 @@ export class UsersController {
     return this.usersService.findAll(paginationQuery);
   }
 
-
-  @UseGuards(JwtAuthGuard)
```
Comment (matboivin) [inline src/nestjs-back/src/users/users.controller.ts:45]: Not deleted, just reorganized the file.

File: src/nestjs-back/src/users/users.controller.ts [hunk 2]
```diff
@@ -61,20 +50,6 @@ export class UsersController {
     });
   }
 
-  /* anyone can create a new user to begin the authentication process */
```
Comment (matboivin) [inline src/nestjs-back/src/users/users.controller.ts:64]: Same. Reorganized the file.

File: src/nestjs-back/src/users/users.service.ts [hunk 3]
```diff
@@ -255,6 +255,23 @@ export class UsersService {
...
+  async updateBlockedUsers(id: string, users: User[]) {
+    const blockList: User[] = [];
+
+    for (var toBlock of users) {
+      if (parseInt(id) !== toBlock.id) { /* User can't block themselve */
+        blockList.push(toBlock);
+      }
+    }
+
+    let user = await this.usersRepository.findOne(id, {
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:267]: You could use `const` here.

File: src/nestjs-back/src/users/users.service.ts [hunk 4]
```diff
@@ -255,6 +255,23 @@ export class UsersService {
...
+      relations: ['blockedUsers'],
+    });
+
+    user.blockedUsers = blockList;
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:271]: Object mutation is possible on a `const` variable. However the variable itself can't be reassigned.

File: src/nestjs-back/src/users/users.service.ts [hunk 5]
```diff
@@ -272,8 +289,9 @@ export class UsersService {
       /* Send frienship invite */
       for (var receiver of updateUserDto.pendingFriendsSent) {
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:290]: const/let instead of var

File: src/nestjs-back/src/users/users.service.ts [hunk 6]
```diff
@@ -419,20 +437,39 @@ export class UsersService {
...
+      throw new Error('Operation not allowed.');
+    }
     let user = await this.usersRepository.findOne(id, {
-      relations: ['pendingFriendsReceived'],
+      relations: ['friends', 'pendingFriendsReceived'],
     });
     const sender = await this.usersRepository.findOne(senderId);
 
     if (!user || !sender) {
       throw new Error('Cannot update user.');
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:449]: more context about the error would be beneficial (user or sender not found)

File: src/nestjs-back/src/users/users.service.ts [hunk 7]
```diff
@@ -419,20 +437,39 @@ export class UsersService {
...
+        user.pendingFriendsReceived.push(sender);
+        return this.usersRepository.save(user);
+      }
+    }
+    return user;
   }
 
   async removeFriendshipReceived(id: string, senderId: string) {
+    if (id === senderId) {
+      throw new Error('Operation not allowed. Please report the bug.');
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:471]: more context would be beneficial

File: src/nestjs-back/src/users/users.service.ts [hunk 8]
```diff
@@ -450,20 +487,51 @@ export class UsersService {
   }
 
   async addFriendshipSent(id: string, receiverId: string) {
+    if (id === receiverId) {
+      throw new Error('Operation not allowed.');
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:491]: more context

File: src/nestjs-back/src/users/users.service.ts [hunk 9]
```diff
@@ -450,20 +487,51 @@ export class UsersService {
...
     let user = await this.usersRepository.findOne(id, {
-      relations: ['pendingFriendsSent'],
+      relations: ['friends', 'pendingFriendsSent'],
+    });
+    const receiver = await this.usersRepository.findOne(receiverId, {
+      relations: [ 'blockedUsers' ]
     });
-    const receiver = await this.usersRepository.findOne(receiverId);
 
     if (!user || !receiver) {
```
Comment (aurelien-brabant) [inline src/nestjs-back/src/users/users.service.ts:500]: context

File: src/nextjs-front/context/relationship/RelationshipProvider.tsx [hunk 10]
```diff
@@ -13,58 +13,65 @@ const RelationshipProvider: React.FC = ({ children }) => {
...
-		const checkSuggested = (list: User[], id: String) => {
-			for (var i in list) {
-				if (list[i].id === id) {
-					return false;
+	/* Set a list of suggested new friends */
+	const createSuggestedFriends = (users: User[], friends: User[], blocked: User[], requests: User[]) => {
+		const suggestedUsers: User[] = [];
+		const userId: string = currentUser.id;
+
+		for (var user of users) {
```
Comment (aurelien-brabant) [inline src/nextjs-front/context/relationship/RelationshipProvider.tsx:21]: const

File: src/nextjs-front/context/relationship/RelationshipProvider.tsx [hunk 11]
```diff
@@ -13,58 +13,65 @@ const RelationshipProvider: React.FC = ({ children }) => {
...
+			if (user.id !== userId && !user.accountDeactivated) {
+				const isNotFriend = !(!!friends.find((friend) => friend.id === user.id));
```
Comment (aurelien-brabant) [inline src/nextjs-front/context/relationship/RelationshipProvider.tsx:23]: `!!` is used to force boolean coercion without reversing the boolean value.
Therefore `!friends.find()` should be enough.

**Review by aurelien-brabant — CHANGES_REQUESTED** [review]
Thank you for such a great PR!
Make sure to provide enough context about the errors leveraged from the service so that we can troubleshoot errors more easily.
Also avoid using `var` which is considered deprecated for several reasons (hoisting being the primary one) and always use `let` or `const` depending on your usage. Most of the time `const` should be enough. If you need to reassign the variable use `let`.

File: src/nestjs-back/src/users/users.service.ts [hunk 12]
```diff
@@ -255,6 +255,23 @@ export class UsersService {
     }
   }
 
+  async updateBlockedUsers(id: string, users: User[]) {
+    const blockList: User[] = [];
+
+    for (var toBlock of users) {
```
[the remark this answers is not in the document]
Comment (matboivin) [reply src/nestjs-back/src/users/users.service.ts:261]: Good to know, ty!

File: src/nestjs-back/src/users/users.service.ts (hunk 4 above)
Comment (matboivin) [reply src/nestjs-back/src/users/users.service.ts:271]: Should I use `let`? Since it `blockedUsers` is indeed updated.

Comment (aurelien-brabant) [reply src/nestjs-back/src/users/users.service.ts:271]: Yeah my bad. Closing that, will fix later.

**Pull request merged**